docs/guides/slowness.md

```diff
@@ -0,0 +1,230 @@
+
+## WebDriverAgent (WDA) Server
+
+WDA source code is located in the separate [repository](https://github.com/appium/WebDriverAgent/tree/master).
+The content of this repository is published as [appium-webdriveragent](https://www.npmjs.com/package/appium-webdriveragent)
+NPM package and contains several helper Node.js modules along with the WDA source code itself.
+This source code is compiled into an .xctrunner bundle, which is a special application type
+that contains tests (also it has some higher privileges in comparison to vanilla apps).
+WebDriverAgent project itself consists of three main parts:
+
+- Vendor Libs
+- WebDriverAgentLib
+- WebDriverAgentRunner
+
+Vendor libs, like RoutingHTTPServer, ensure the support for low-level HTTP- and TCP-server APIs.
+WebDriverAgentLib defines handlers for [W3C WebDriver](https://www.w3.org/TR/webdriver/) endpoints
+and implements all the heavy-lifting procedures related to Apple's XCTest communication
+and some more custom stuff specific for the XCUITest driver.
+WebDriverAgentRunner is actually one long test, whose main purpose
+is to run the HTTP server implemented by the WebDriverAgentLib.
+
+Important conclusions from the above information:
+
+- WDA is an HTTP server, which executes API commands by invoking HTTP response handlers
+- WDA uses Apple's XCTest APIs with various custom additions
+
+## How to confirm my script's bottleneck is WDA
+
+Check the server logs in order to verify how long it takes for the XCUITest driver to receive a
+response from WDA. The log line that is written before an HTTP request is proxied to WDA looks
+like `Proxying [X] to [Y]`. Also consider enabling server timestamps by providing the
+`--log-timestamp` command line parameter. If you observe timestamps between the above log line and the
+next one differ too much and the difference is an anomaly (e.g. the same step is (much) faster
+for other apps/environments/parameter combinations) then it might serve as a confirmation of a
+suspicious slowness.
+
+## Patterns lookup
+
+After the slowness is confirmed it is important to determine behavior patterns, e.g. under which
+circumstances does it happen, if it is always reproducible, etc. This article only targets specific
+patterns that the author knows of or dealt with. If your pattern is not present here then try to
+look for possible occurrences in existing [issues](https://github.com/appium/appium/issues),
+[Appium forum](https://discuss.appium.io) or just search the internet.
+
+## Pattern: Application startup is slow
+
+### Symptoms
+
+You observe timeouts or unusual slowness (in comparison to manual execution performance)
+of the application startup on session init (if it also includes app startup)
+or mid-session app startup.
+
+### Causes
+
+When XCTest starts an app it ensures the accessibility layer of it is ready for interactions.
+To check that the framework verifies the application is idling (e.g. does perform any actions
```

Comment:
I assume this is meant to be 'does not perform any actions'

Reply:
changed, good catch